chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,185 @@
+package mir
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/types"
+	ltypes "github.com/filecoin-project/lotus/chain/types"
+	mirTypes "github.com/filecoin-project/mir/pkg/types"
+)
+
+// Mine mines Filecoin blocks.
+
+// Mine implements the following algorithm:
+// 1. Retrieve messages and cross-messages from mempool.
+// 2. Send these messages to the Mir node.
+// 3. Receive ordered messages from the Mir node and push them into the next Filecoin block.
+// 4. Submit this block.
+func Mine(ctx context.Context, miner address.Address, api v1api.FullNode) error {
+	log.Info("Mir miner started")
+	defer log.Info("Mir miner stopped")
+
+	// TODO: Suppose we want to use Mir in Root and in a subnet at the same time.
+	// Do we need two Mir agents for that?
```

Comment:
Probably? I would have to look deeper on how Mir is implemented, but probably that is the cleanest approach. Probably @matejpavlovic can clarify this one.

Reply:
Yes I think that is the way to do it. Unless we want to somehow explicitly share the consensus mechanism for both the root and the subnet (in which case it's questionable why we even have a subnet), there should be two distinct mir agents.